File: leetcode/Medium/1438 Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/1438.java
class Solution {
    public int longestSubarray(int[] nums, int limit) {
        int l = 0, length = 0;
        TreeMap<Integer, Integer> map = new TreeMap<>();
        for(int i=0; i<nums.length; i++){
            map.put(nums[i], map.getOrDefault(nums[i], 0) + 1);

            if(map.lastKey() - map.firstKey() <= limit){
                length = Math.max(length, i-l+1);
            }
            while(l < i && map.lastKey() - map.firstKey() > limit){
                map.put(nums[l], map.get(nums[l])-1);
                if(map.get(nums[l]) == 0){
                    map.remove(nums[l]);
                }
                l++;
            }
            
        }


        return length;
    }
}

// TC: O(nlog(n))
// SC: O(n)

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> maxd = new ArrayDeque<>();
        Deque<Integer> mind = new ArrayDeque<>();
        int l=0, i=0;
        for(;i<nums.length; i++){
            while(!maxd.isEmpty() && nums[i] > maxd.peekLast()) maxd.pollLast();
            while(!mind.isEmpty() && nums[i] < mind.peekLast()) mind.pollLast();
            maxd.add(nums[i]);
            mind.add(nums[i]);
            if(maxd.peek() - mind.peek() > limit){
                if(maxd.peek() == nums[l]) maxd.poll();
                if(mind.peek() == nums[l]) mind.poll();
                l++;
            }
        }
        return i-l;
    }
}

// TC: O(n)
// SC: O(n)

File: leetcode/Medium/1438 Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/1438.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        int l = 0, res = 0;
        multiset<int> ms;

        for(int i=0; i<nums.size(); i++){
            ms.insert(nums[i]);

            if(*ms.rbegin() - *ms.begin() <= limit)
                res = max(res, i-l+1);

            while(l < i && *ms.rbegin() - *ms.begin() > limit){
                ms.erase(ms.find(nums[l++]));
            }
            
        }

        return res;
    }
};


// TC: O(nlog(n))
// SC: O(n)


class Solution {
public:
    int longestSubarray(vector<int>& nums, int limit) {
        deque<int> maxd, mind;
        int l=0, i=0;
        for(;i<nums.size(); i++){
            while(!maxd.empty() && nums[i] > maxd.back()) maxd.pop_back();
            while(!mind.empty() && nums[i] < mind.back()) mind.pop_back();
            maxd.push_back(nums[i]);
            mind.push_back(nums[i]);

            if(maxd.front() - mind.front() > limit){
                if(maxd.front() == nums[l]) maxd.pop_front();
                if(mind.front() == nums[l]) mind.pop_front();
                l++;
            }
        }


        return i-l;
    }
};

// TC: O(n)
// SC: O(n)
